Fix handling of validations that require multiple feature gates

JoelSpeed opened the pull request:
We have a couple of examples (for example https://github.com/openshift/api/pull/2131) of validations that end up relying on multiple feature gates. Thus far, this has caused an issue where one gate is promoted, but the other is not, since the validation cannot be automatically handled if the gated validation is promoted prior to the ungated dependent gate.

This PR updates the example API with a second Example2 feature gate, and then adds a validation that is only enabled when both the Example and Example2 feature gates are enabled (as can be seen by the fact that the validation is enabled in Custom and Dev CRDs, but not TPNU where Example2 is disabled.

I need to merge https://github.com/openshift/kubernetes-sigs-controller-tools/pull/25 and update the vendor here before we can merge this

CC @gcs278 @deads2k

JoelSpeed commented:
/approve cancel

@deads2k needs to approve this one

deads2k commented on features/legacyfeaturegates.go:
```diff
@@ -37,6 +37,8 @@ var legacyFeatureGates = sets.New(
 	// never add to this list, if you think you have an exception ask @deads2k
 	"Example",
 	// never add to this list, if you think you have an exception ask @deads2k
+	"Example2",
+	// never add to this list, if you think you have an exception ask @deads2k
```
I accept the need for this

deads2k commented on features/features.go:
```diff
@@ -490,6 +490,14 @@ var (
 
+	FeatureGateExample2 = newFeatureGate("Example2").
+				reportProblemsToJiraComponent("cluster-config").
+				contactPerson("deads").
```
naw, this joy is yours.

JoelSpeed replied:
That is entirely fair

deads2k commented on tools/go.mod:
```diff
@@ -34,7 +34,7 @@ require (
 )
 
-replace sigs.k8s.io/controller-tools => github.com/openshift/controller-tools v0.12.1-0.20241105174925-ead1ef573704
+replace sigs.k8s.io/controller-tools => github.com/JoelSpeed/controller-tools v0.6.3-0.20250106114547-5b11e5175d3d
```
temporary?

JoelSpeed replied:
Yeah, need to merge the tools PR first and then will get this fixed up

JoelSpeed commented on example/v1/types_stable.go:
```diff
@@ -33,6 +33,7 @@ type StableConfigType struct {
 
 // StableConfigTypeSpec is the desired state
 // +openshift:validation:FeatureGateAwareXValidation:featureGate=Example,rule="has(oldSelf.coolNewField) ? has(self.coolNewField) : true",message="coolNewField may not be removed once set"
+// +openshift:validation:FeatureGateAwareXValidation:requiredFeatureGate=Example;Example2,rule="has(oldSelf.stableField) ? has(self.stableField) : true",message="stableField may not be removed once set (this should only show up with both the Example and Example2 feature gates)"
```
Yep, `;` is the standard separator, you can see in the code changes where we split this, that you're already splitting on `;` for the other gate splitting

It's also used commonly for `kubebuilder:validation:Enum:=A;B;C`

gcs278 commented on example/v1/zz_generated.featuregated-crd-manifests.yaml:
```diff
@@ -6,6 +6,7 @@ stableconfigtypes.example.openshift.io:
   Category: ""
   FeatureGates:
   - Example
+  - Example+Example2
```
If I understand this correctly, this change creates "compound" or "composite" feature gates for markers that use > 1 feature gate, like `SetEIPForNLBIngressController+IngressControllerLBSubnetsAWS` for https://github.com/openshift/api/pull/2131.

Does that impact the promotion to defaulting (GA) process? Will `features/features.go` require us to specify the composite feature gates when we promote them too? Or if you promote both Example and Example2, then the composite Example+Example2 gets promoted? Or is the promote-to-default process not required for Example+Example2?

JoelSpeed replied:
> Does that impact the promotion to defaulting (GA) process?

No, it only affects the generation of what we refer to as partial schemas.

Each API here has 4 versions of the CRD, Custom, DevPreview, TechPreview and Stable.

These CRDs are merged from partial schemas, which take the base level (Stable) and then add feature gates and the relevant schema changes for those feature gates.

Prior to this change, those partial schemas always only included one gate. With this change, some of those partial schemas now include more than one gate, where a particular validation requires more than one gate.

To create the custom/dp/tp/stable versions, we merge the base CRD, with all of the gated versions, for the gates enabled within that set.

If you look at the difference between DP and TP in this example, you'll see that the validation gated on Example+Example2, is only present in the DP version of the CRD, because while the Example gate is enable in DP and TP, the Example2 gate is only present in the DP version.

These composite gates aren't real per se, and are only used in parts of the generation. You'd never see a composite gate in the cluster or in any of the generated manifests that make it into the payload

JoelSpeed commented:
Addressed review comments, fixed up temporary import to a separate branch, this is now ready to go

Taking https://github.com/openshift/api/pull/2133#discussion_r1904199126 as David's approval of new gate

/approve

gcs278 commented:
@JoelSpeed friendly ping on the verify failure:
```
No test suite file found for CRD /go/src/github.com/openshift/api/example/v1/zz_generated.featuregated-crd-manifests/stableconfigtypes.example.openshift.io/Example+Example2.yaml, expected /go/src/github.com/openshift/api/example/v1/tests/stableconfigtypes.example.openshift.io/Example+Example2.yaml
```
My PR https://github.com/openshift/api/pull/2131 is blocked on this

JoelSpeed commented:
/retest-required
